### PR TITLE
add disable workflow feature

### PR DESCRIPTION
**Pull request by a1012112796**
As title, that's simmilar with github.

image

image

### CODE REVIEW

File: models/repo/repo_unit.go [hunk 1]
```diff
@@ -162,6 +163,59 @@ func (cfg *PullRequestsConfig) GetDefaultMergeStyle() MergeStyle {
...
+}
+
+func (cfg *ActionsConfig) ToString() string {
+	result := ""
+
+	for _, workflow := range cfg.DisabledWorkflows {
+		result += workflow + ","
+	}
+
+	return strings.TrimSuffix(result, ",")
```
Comment (wolfogre) [inline models/repo/repo_unit.go]: Why not

```suggestion
	return strings.Join(cfg.DisabledWorkflows, ",")
```

File: routers/web/repo/actions/view.go [hunk 2]
```diff
@@ -572,3 +573,43 @@ func ArtifactsDownloadView(ctx *context_module.Context) {
...
+
+func DisableWorkflowFile(ctx *context_module.Context) {
+	disableOrEnableWorkflowFile(false, ctx)
+}
+
+func EnableWorkflowFile(ctx *context_module.Context) {
+	disableOrEnableWorkflowFile(true, ctx)
+}
+
+func disableOrEnableWorkflowFile(isEnable bool, ctx *context_module.Context) {
```
Comment (wolfogre) [inline routers/web/repo/actions/view.go]: I think it's a good idea to keep the ctx as the first parameter.
```suggestion
func disableOrEnableWorkflowFile(ctx *context_module.Context, isEnable bool) {
```

**Comment by yp05327** [comment]
How about only add a toggle switch here:
image
Instead of Enable/Disable button here, it seems not clear which workflow to Enable/Disable and you need to select this workflow in the list first.
image

**Comment by silverwind** [comment]
UI might look a bit cramped with switches in the workflow list, and I do see a certain benefit of having the button in the same place as GH.

**Review by wolfogre — APPROVED** [review]
LGTM on the backend part.

**Comment by silverwind** [comment]
Did some UI enhancements. Menu button that submits a hidden form, red label (imho warranted to have this call to attention via color):

Screenshot 2023-08-11 at 08 07 08
Screenshot 2023-08-11 at 08 07 19

I noticed the flashes sent by the backend do not work for me, even before my changes, so I those need to be fixed before landing.

File: models/repo/repo_unit.go [hunk 3]
```diff
@@ -162,6 +163,53 @@ func (cfg *PullRequestsConfig) GetDefaultMergeStyle() MergeStyle {
 	return MergeStyleMerge
 }
 
+type ActionsConfig struct {
+	DisabledWorkflows []string
+}
+
+func (cfg *ActionsConfig) EnableWrokflow(file string) {
```
Comment (silverwind) [inline models/repo/repo_unit.go]: Typo `Wrokflow`, please fix in all function names.

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -3488,6 +3491,11 @@ runs.status_no_select = All status
 runs.no_results = No results matched.
 runs.no_runs = The workflow has no runs yet.
 
+workflow.disable = Disable Workflow
+workflow.disable_success = The workflow '%s' has been disabled successfully!
```
Comment (silverwind) [inline options/locale/locale_en-US.ini]:
```suggestion
workflow.disable_success = Workflow '%s' disabled successfully.
```

File: options/locale/locale_en-US.ini [hunk 5]
```diff
@@ -3488,6 +3491,11 @@ runs.status_no_select = All status
...
+workflow.enable = Enable Workflow
+workflow.enable_success = The workflow '%s' has been enabled successfully!
```
Comment (silverwind) [inline options/locale/locale_en-US.ini]:
```suggestion
workflow.enable_success = Workflow '%s' enabled successfully.
```

**Comment by silverwind** [comment]
Flashes need to be fixed (ideally) or removed, then it's LGTM.

File: models/repo/repo_unit.go [hunk 6]
```diff
@@ -162,6 +163,53 @@ func (cfg *PullRequestsConfig) GetDefaultMergeStyle() MergeStyle {
...
+	DisabledWorkflows []string
+}
+
+func (cfg *ActionsConfig) EnableWorkflow(file string) {
+	for i, workflow := range cfg.DisabledWorkflows {
+		if file == workflow {
+			cfg.DisabledWorkflows = append(cfg.DisabledWorkflows[:i], cfg.DisabledWorkflows[i+1:]...)
+			break
+		}
+	}
```
Comment (wxiaoguang) [inline models/repo/repo_unit.go]: `util.SliceRemoveAll`?

File: models/repo/repo_unit.go [hunk 7]
```diff
@@ -162,6 +163,53 @@ func (cfg *PullRequestsConfig) GetDefaultMergeStyle() MergeStyle {
...
+}
+
+func (cfg *ActionsConfig) IsWorkflowDisabled(file string) bool {
+	for _, workflow := range cfg.DisabledWorkflows {
+		if file == workflow {
+			return true
+		}
+	}
+
+	return false
```
Comment (wxiaoguang) [inline models/repo/repo_unit.go]: `util.SliceContains`?

File: routers/web/repo/actions/view.go [hunk 8]
```diff
@@ -572,3 +573,43 @@ func ArtifactsDownloadView(ctx *context_module.Context) {
...
+	}
+
+	if isEnable {
+		ctx.Flash.Success(ctx.Tr("actions.workflow.enable_success", workflow))
+	} else {
+		ctx.Flash.Success(ctx.Tr("actions.workflow.disable_success", workflow))
+	}
+
+	redirectURL := fmt.Sprintf("%s/actions?workflow=%s&actor=%s&status=%s", ctx.Repo.RepoLink, workflow,
+		ctx.FormString("actor"), ctx.FormString("status"))
```
Comment (wxiaoguang) [inline routers/web/repo/actions/view.go]: Please escape every parameter for the URL

File: templates/repo/actions/list.tmpl [hunk 9]
```diff
@@ -13,12 +13,16 @@
...
+							{{if $.ActionsConfig.IsWorkflowDisabled .Entry.Name}}
+								<div class="ui red label">{{$.locale.Tr "disabled"}}</div>
+							{{end}}
 						</a>
 					{{end}}
 				</div>
 			</div>
 			<div class="twelve wide column content">
-				<div class="ui secondary filter stackable menu gt-je">
+				<div class="ui secondary filter stackable menu gt-je gt-df gt-ac">
```
Comment (wxiaoguang) [inline templates/repo/actions/list.tmpl]: IIRC `flex` conflicts with `stackable menu`?

Have it been tested with small width screen?

File: templates/repo/actions/list.tmpl [hunk 10]
```diff
@@ -57,6 +61,18 @@
 							{{end}}
 						</div>
 					</div>
+
+					{{if .AllowDisableOrEnableWorkflow}}
+						<button class="ui small compact jump dropdown icon btn interact-bg gt-p-3">
```
Comment (wxiaoguang) [inline templates/repo/actions/list.tmpl]: `ui dropdown button` is good enough, I think there is no need to use `btn interact-bg gt-p-3`.

File: web_src/js/features/repo-actions.js [hunk 11]
```diff
@@ -0,0 +1,8 @@
+export function initRepoActionList() {
+  const disableWorkflowBtn = document.getElementById('disable-workflow-btn');
+  if (disableWorkflowBtn) {
+    disableWorkflowBtn.addEventListener('click', () => {
+      document.getElementById('disable-workflow-form').submit();
+    });
+  }
```
Comment (wxiaoguang) [inline web_src/js/features/repo-actions.js]: No need, please use `link-action`

**Comment by silverwind** [comment]
Regarding flash, I'm pretty sure we are missing the flash subtemplate somewhere on this page. We should refactor the templates to ensure this subtemplate is included in every page.

File: templates/repo/actions/list.tmpl (hunk 10 above)
Comment (silverwind) [reply templates/repo/actions/list.tmpl]: Such a button would not fit visually:

image

Comment (silverwind) [reply templates/repo/actions/list.tmpl]: I removed some unnecessary classes in https://github.com/go-gitea/gitea/pull/26413/commits/b2a5c22c88be5de5675a9460375ac520f648b87b, I think this is the minimum to achieve the desired look and function.

File: web_src/js/features/repo-actions.js (hunk 11 above)
Comment (silverwind) [reply web_src/js/features/repo-actions.js]: Got an example?

File: templates/repo/actions/list.tmpl (hunk 9 above)
Comment (silverwind) [reply templates/repo/actions/list.tmpl]: Hmm yes, does not look ideal

image

Comment (silverwind) [reply templates/repo/actions/list.tmpl]: Fixed, see below.

**Comment by silverwind** [comment]
Mobile view fixed:

Screenshot 2023-08-12 at 12 21 24

There was a small issue that the margin was too little caused by Fomantic's negative margins on `.ui.secondary.menu`. I have set these to zero now and checked various pages and it seems fine. On issue list, removing the negative margin also fixes this misalignment:

Before:
Screenshot 2023-08-12 at 12 17 39

After:
Screenshot 2023-08-12 at 12 17 32

File: web_src/js/features/repo-actions.js (hunk 11 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-actions.js]: The examples on the devtest page:

http://localhost:3000/devtest/fetch-action

image

File: web_src/css/base.css [hunk 12]
```diff
@@ -653,6 +653,11 @@ a.label,
   color: var(--color-text);
 }
 
+.ui.secondary.menu {
+  margin-left: 0;
+  margin-right: 0;
+}
+
```
Comment (wxiaoguang) [inline web_src/css/base.css]: I do not think it's safe to do so (for a general style)

IIRC, the Fomantic uses margin/padding tricks to align container/children elements. The margins are necessary for using Fomantic components correctly.

Comment (silverwind) [reply web_src/css/base.css]: I thought so too, but after some testing, I could not notice any negative effects of doing this.

File: web_src/js/features/repo-actions.js (hunk 11 above)
Comment (silverwind) [reply web_src/js/features/repo-actions.js]: Was fixed in https://github.com/go-gitea/gitea/pull/26413/commits/70d758b2ae01f12812fc772e2e15cc60ddea2caa.

File: web_src/css/base.css (hunk 12 above)
Comment (silverwind) [reply web_src/css/base.css]: The negative margins was only there to compensate the menu items's margin on first and last item:

```css
.ui.secondary.menu .item {
  margin: 0 0.35714286em;
}
```

This was it:

Screenshot 2023-08-14 at 14 31 19

In https://github.com/go-gitea/gitea/pull/26413/commits/0ac99a21422264451df4375e1e9f4b529cdc5989, I have now replaced this mechanism with `gap`, which makes the negative margin unnecessary and this is the result:

Screenshot 2023-08-14 at 14 35 25

**Review by silverwind — APPROVED** [review]
Looks good, waiting on @wxiaoguang for https://github.com/go-gitea/gitea/pull/26413#discussion_r1293415827.

**Comment by wxiaoguang** [comment]
> Looks good, waiting on @wxiaoguang for [#26413 (comment)](https://github.com/go-gitea/gitea/pull/26413#discussion_r1293415827).

Although I haven't tested nor fully understood, I think it might be not bad, so no objection.

**Comment by silverwind** [comment]
I have tested it across many pages, no issues with the menu changes, so I'm pretty sure they are safe.

**Pull request merged**